pkg/suggestion/v1alpha3/goptuna/service.go

```diff
@@ -0,0 +1,129 @@
+package suggestion_goptuna_v1alpha3
+
+import (
+	"context"
+
+	"github.com/c-bata/goptuna"
+	"github.com/kubeflow/katib/pkg/apis/manager/v1alpha3"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"k8s.io/klog"
+)
+
+const (
+	AlgorithmCMAES  = "cmaes"
+	AlgorithmTPE    = "tpe"
+	AlgorithmRandom = "random"
+)
+
+func NewSuggestionService() *SuggestionService {
+	return &SuggestionService{}
+}
+
+type SuggestionService struct{}
+
+func (s *SuggestionService) GetSuggestions(
+	ctx context.Context,
+	req *api_v1_alpha3.GetSuggestionsRequest,
+) (*api_v1_alpha3.GetSuggestionsReply, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "request is empty")
+	}
+
+	direction, err := toGoptunaDirection(req.GetExperiment().GetSpec().GetObjective().GetType())
+	if err != nil {
+		klog.Errorf("Failed to convert to Goptuna direction: %s", err)
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	independentSampler, relativeSampler, err := toGoptunaSampler(req.GetExperiment().GetSpec().GetAlgorithm())
+	if err != nil {
+		klog.Errorf("Failed to create Goptuna sampler: %s", err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	searchSpace, err := toGoptunaSearchSpace(req.GetExperiment().GetSpec().GetParameterSpecs().GetParameters())
+	if err != nil {
+		klog.Errorf("Failed to convert to Goptuna search space: %s", err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	klog.Infof("Goptuna search space: %#v", searchSpace)
+
+	studyOpts := make([]goptuna.StudyOption, 0, 3)
+	studyOpts = append(studyOpts, goptuna.StudyOptionSetDirection(direction))
+	if independentSampler != nil {
+		studyOpts = append(studyOpts, goptuna.StudyOptionSampler(independentSampler))
+	}
+	if relativeSampler != nil {
+		studyOpts = append(studyOpts, goptuna.StudyOptionRelativeSampler(relativeSampler))
+	}
+
+	study, err := goptuna.CreateStudy("katib", studyOpts...)
+	if err != nil {
+		klog.Errorf("Failed to create Goptuna study: %s", err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	trials, err := toGoptunaTrials(req.GetTrials(), study, searchSpace)
+	if err != nil {
+		klog.Errorf("Failed to convert to Goptuna trials: %s", err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	for _, t := range trials {
+		_, err = study.Storage.CloneTrial(study.ID, t)
```

Comment:
You will override Trial information each `GetSuggestion` call. Is it fine for Goptuna ?

Reply:
Thank you for your comment. I found a bug that heavily affected optimization performance when using CMA-ES.

Reply:
I fixed this bug at 4522744b.

| Before | After |
| --- | --- |
| ![ScreenShot 2020-04-09 18 21 56](https://user-images.githubusercontent.com/5564044/78879562-03bf4480-7a8f-11ea-9037-ee3470919617.png) | <img width="1482" alt="ScreenShot 2020-04-11 9 01 43" src="https://user-images.githubusercontent.com/5564044/79030266-22c6ef00-7bd3-11ea-8763-7da1b660241f.png"> |